src/Microsoft.DotNet.Build.VSTSBuildsApi/VSTSBuildClient.cs

```diff
@@ -0,0 +1,228 @@
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="definition">VSTS build definition JSON object</param>
+        /// <returns>Created or updated build definition id</returns>
+        private string CreateOrUpdateDefinition(JObject definition)
+        {
+            var key = string.Join("_", _collectionIdentifier, definition["name"].ToString());
+            definition["name"] = key;
+            var vstsDefinitions = VSTS_RetrieveDefinitionsListByNameAndPath(definition);
+
+
+            if (vstsDefinitions.Count == 0)
+            {
+                /* Create */
+                VSTS_CreateDefinition(definition);
+                definition = VSTS_RetrieveDefinitionById(definition);
+            }
+            else if(vstsDefinitions.Count == 1)
+            {
+                var vstsDefinition = vstsDefinitions[0];
+
+                /* Update */
+                if (!IsSameContents(definition, vstsDefinition))
+                {
+                    definition["revision"] = vstsDefinition["revision"];
+                    definition["id"] = vstsDefinition["id"];
+                    definition["uri"] = vstsDefinition["uri"];
+                    definition["path"] = vstsDefinition["path"];
+                    definition["createdDate"] = vstsDefinition["createdDate"];
+                    definition["url"] = vstsDefinition["url"];
+                    definition["comment"] = "Automated update";
+
+                    VSTS_UpdateDefinition(definition);
+                }
+            }
+            else
+            {
+                throw new Exception(string.Format("Obtained multiple {0} definitions with the same 'name' ({1}) and 'path' ({2}) properties.  This should not be possible.", vstsDefinitions.Count, vstsDefinitions[0]["name"].ToString(), vstsDefinitions[0]["path"].ToString()));
+            }
+            return definition["id"].ToString();
+        }
+
+        private JObject GetDefinition(Stream stream)
+        {
+            using (StreamReader streamReader = new StreamReader(stream))
+            using (JsonTextReader jsonReader = new JsonTextReader(streamReader))
+            {
+               return (JObject)JToken.ReadFrom(jsonReader);
+            }
+        }
+
+        /// <summary>
+        /// Update VSTS Build Definition
+        /// </summary>
+        /// <param name="definition"></param>
+        private void VSTS_UpdateDefinition(JObject definition)
+        {
+            string requestUri = $"DefaultCollection/DevDiv/_apis/build/definitions/{definition["id"].ToString()}?api-version=2.0";
+
+            HttpContent content = new StringContent(JsonConvert.SerializeObject(definition), System.Text.Encoding.UTF8, "application/json");
+            ProcessResponseStatusCode(GetClient(definition["url"].ToString()).PutAsync(requestUri, content).Result);
```

Comment:
PipeBuild wasn't designed with `async` in mind, and with the current implementation, I don't want to rewrite a bunch of stuff (in PipeBuild) to support it, but I agree with the principle, and some other tool may want to take advantage of it.  I'll make that change.